### ISSUE TITLE
Display races in a taxonomy tree in the help

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5093]
In the help, we could use a tree style to organise the races in the "Units" section, so that the list doesn't become too long.

* The Dunefolk could be included in Humans.
* The Quenoth could be included in Elves (see #4876), as could many of the races from Ageless Era.
* The falcons, gryphons, bats, etc could be moved in to "monsters" (see #5039, copied below)

For configuring this, I suggest we add `[race]taxonomy=`, which would allow add-ons that add races to also define where those races go. For example,

    [race]
        plural_name="race^Desert Elves"
        taxonomy=elves
        ...

Posted by @soliton- in https://github.com/wesnoth/wesnoth/pull/5039#issuecomment-674541396 :

> If the list is too crowded now then perhaps propose to group some races together. As @CelticMinstrel said there can potentially be broader race definitions. Otherwise I don't think there is a real solution unless we just want to have less unit types or somehow make up a taxonomy tree...

> An arbitrary split of the catch-all monster race we have now into multiple generic races like "animal" or "creature" does not seem useful. It'll just be confusing to try to remember if some unit was a monster or animal or creature. (And if there is no reason to remember that then why do the split in the first place.)

That second paragraph was replied to by @doofus-01 in https://github.com/wesnoth/wesnoth/pull/5039#issuecomment-674581735 :

> Falcons, gryphons, bats, and ogres have their own race already, seems we've made our decision.  I guess they can just go into monsters if they are solo units (no variations or other levels), otherwise, they get a unique race.

**Comment by stevecotton** [comment]
> How would you solve the other problem of giving them different lore in the help pages?

I was going to have a separate `[race]id=quenoth`, but need to reconsider that based on this discussion. @shikadiqueen , which solution would you use?

**Comment by CelticMinstrel** [comment]
My question was aimed more at @shikadiqueen I guess, but…

What about race inheritance?

```
[race]
    id=quenoth
    parent=elves # whatever you want to call the key
    # unique name/lore text or whatever
[/race]
```

Or something like that? Would that be a reasonable solution?

**Comment by shikadiqueen** [comment]
> How to deal with "race" descriptions is a different deal entirely, though.

If I don't propose a solution for an issue I'm pointing out, there's a good chance I don't have any ideas. 😛

**Comment by shikadiqueen** [comment]
I will say, though, that implementing another "inheritance" type of system like the ones we already have (`[unit_type]` vs `[variation]`, `[male]`, `[female]`, `[base_unit]`) is a terrible idea, because:

1. Content authors are not programmers and inheritance makes for code that's harder to follow even for veteran programmers.
2. Dealing with more than one level of inheritance in tooling (`wmlunits`, etc.) and documentation is a nightmare.
3. I think `[unit_type]`, `[variation]`, `[male]`, `[female]`, `[base_unit]`, `[unit]` and `[modifications]` constantly haunting me in my nightmares is more than enough, thank you.

**Comment by stevecotton** [comment]
I agree about avoiding yet another inheritance system; I'm thinking of calling it `help_taxonomy` and limiting it to just this:

`[race]help_taxonomy=`: `{{DevFeature1.15|?}}` in the help browser, show this race's units as a subfolder of another race; the value of this attribute should be the other race's `id` attribute. This only affects the help browser, for all other purposes (such as WML filters) the two races are completely separate.

**Comment by CelticMinstrel** [comment]
So then, elves and quenoth would still be separate races, as would humans and dunefolk?

**Comment by stevecotton** [comment]
Yes

**Comment by gfgtdf** [comment]
I'm not really oncinved of this idea, I think the cases of 'too many races' is quite a corner case (okay i know that most mp people have ageless installed but still) and in is quite likely not to be fixed by this because if two add-ons create two similar races they don't know each other so none of these would get the idea to be a   subtype of each other.

Furthermore i think it makes races harder to find rather than easier since  the decision when a race if a subtype of another will probably just as abritary as before (are drakes a subtype of monsters? are units where a human rides on an animal humans, animal or neither? and most impotantly: does the autor of the addon use the taxonomy feature or not). So while before it was enough to know the name of a race, of easily find it in the (alphabetically sorted) list they'd then have to guess of what taxonomy the race are probably search quite a bit.

**Comment by CelticMinstrel** [comment]
Is there a search feature in help? Most of the time the taxonomy would be obvious, but in case it's not, a search feature would solve the issue of people not knowing where to look. Or even just a filter on the topic tree.

**Comment by doofus-01** [comment]
> I'm not really oncinved of this idea, I think the cases of 'too many races' is quite a corner case (okay i know that most mp people have ageless installed but still)

Too many races isn't the only problem this is trying to solve, is it?  "Why aren't Dunefolk considered human?" is also a big part of it. 

> Is there a search feature in help? 

There doesn't seem to be, and the layout of such a feature would take some thought, since "units" and "terrain" and "commands" and all else are in the tree.

Two thoughts on how to avoid some of the issues Iris mentioned:
1.  Would it make sense to move all the lore-specific stuff (such as the "Subject of the Crown" and "The Clansmen" in the current human description) to an "Encyclopedia" entry, but have links to the encyclopedia in the race description?  

2.  If it was possible (but not mandatory) to insert a link to a certain race that allowed it (humans, for example) from another race (e.g. dunefolk), that might make it easier for someone reading the Help to find what they are looking for if the tree view on the left is too confusing, but not require another inheritance.

**Comment by CelticMinstrel** [comment]
If we could just switch over to the GUI2 help view, I believe a filter of the topic tree would be pretty easy… but that said, I thought there actually was a search feature in the GUI1 help? Am I imagining this?

I'm pretty sure it should be possible to insert a link from the Dunefolk help page to the Humans help page. Just add some code similar to `<ref text='Humans' target='humans'>`. I don't remember the exact key names though, nor do I know the correct format for the topic ID of a race topic. For the former there's multiple examples in other help topics.

**Comment by doofus-01** [comment]
> I'm pretty sure it should be possible to insert a link from the Dunefolk help page to the Humans help page. 

I meant the other way: get the Dunefolk config to insert a link about Dunefolk into the generic Humans page.  So that the generic Humans configuration doesn't have to know anything about the Dunefolk ahead of time.

**Comment by CelticMinstrel** [comment]
Ah, I see. Well, given the `help_taxonomy` key that was proposed, I think this should be relatively simple to do.

**Comment by stevecotton** [comment]
I've opened PR #5153 for having links between races.

I'm wondering about also adding a way to add a single unit to multiple races, which I'd add as a `[unit_type]help_taxonomy=`, attribute. For example
```
[unit_type]
    id=Goblin Knight
    race=wolf
    help_taxonomy=goblin
    ...
```

I've abandoned my initial idea of tree-style subfolders - once I saw it, I didn't like the look of it. My prototype for it is below, note the "todo" in it still needs fixing (it worked for Quenoth to Elves, but broke for Dunefolk to Humans).
![wesnoth_help_taxo_treeview](https://user-images.githubusercontent.com/101462/92568621-b9313680-f27f-11ea-96ea-156b40e11e78.png)

```
diff --git a/src/help/help_impl.cpp b/src/help/help_impl.cpp
index 2ac216d2ac9..f2ce4d8ebe6 100644
--- a/src/help/help_impl.cpp
+++ b/src/help/help_impl.cpp
@@ -786,27 +786,42 @@ void generate_races_sections(const config *help_cfg, section &sec, int level)
                }
        }
 
-       for(std::set<std::string, string_less>::iterator it = races.begin(); it != races.end(); ++it) {
+       for(const auto race_id : races) {
                section race_section;
                config section_cfg;
 
-               bool hidden = (visible_races.count(*it) == 0);
+               bool hidden = (visible_races.count(race_id) == 0);
 
-               section_cfg["id"] = hidden_symbol(hidden) + race_prefix + *it;
+               section_cfg["id"] = hidden_symbol(hidden) + race_prefix + race_id;
 
                std::string title;
-               if (const unit_race *r = unit_types.find_race(*it)) {
+               std::string help_taxonomy;
+               if (const unit_race *r = unit_types.find_race(race_id)) {
                        title = r->plural_name();
+                       help_taxonomy = r->help_taxonomy();
                } else {
                        title = _ ("race^Miscellaneous");
+                       // leave help_taxonomy empty
                }
                section_cfg["title"] = title;
 
-               section_cfg["sections_generator"] = "units:" + *it;
-               section_cfg["generator"] = "units:" + *it;
+               section_cfg["sections_generator"] = "units:" + race_id;
+               section_cfg["generator"] = "units:" + race_id;
 
                parse_config_internal(help_cfg, &section_cfg, race_section, level+1);
-               sec.add_section(race_section);
+
+               // Support [race]help_taxonomy=. This implementation assumes that the broader
+               // race has been added first, but has a fallback to just adding the new race
+               // at the top level, as if there wasn't a help_taxonomy declaration.
+               // \todo: clear up that assumption
+               section *parent = nullptr;
+               if (!help_taxonomy.empty()) {
+                       parent = find_section(sec, hidden_symbol(hidden) + race_prefix + help_taxonomy);
+               }
+               if (!parent) {
+                       parent = &sec;
+               }
+               parent->add_section(race_section);
        }
 }
 
```

**Comment by stevecotton** [comment]
The code prototyped in the previous comment is now PR #7228.

**Comment by soliton-** [comment]
Should this still be open?

**Comment by stevecotton** [comment]
I can't remember why I left it open after #7228. I assume there was something still to do but I haven't yet thought about what. I'm happy for someone to close it.
